Resolve config when it's a function

piotr-cz opened the pull request:
Rollup accepts function that returns object/ array as config: [docs: Configuration files](https://rollupjs.org/guide/en/#configuration-files).

This is particularly handy when working with async function to prepare configuration

```js
export default (async (commandOptions) => {
  // Some async stuff

  return [
    {
        input: 'foobar.js'
        // ...
    }
  ]
})
```

However this doesn't work with nollup, giving `UnhandledPromiseRejectionWarning: Error: Input option not defined`.
This PR fixes it, however unlike rollup, function is run without command line arguments.

Related rollup code: <https://github.com/rollup/rollup/blob/v2.21.0/cli/run/loadConfigFile.ts#L113-L125>

piotr-cz commented:
ugh, this seems to break Hot Module Replacement

PepsRyuu commented:
This could end up having a huge backwards breaking impact on users from the looks of it.

Avoid using ```async/await```, that'll help quite a bit. :)

piotr-cz commented:
This adds support to use function/ promise to object/ array - so there _should not_ be B/C issue.

However I might not have full picture, as it breaks all Dev Middleware tests and simply adding await in tests doesn't help

PepsRyuu commented:
Changes the function signature. Anyone using the dev server or middleware will have to update their code to put in an await.

piotr-cz commented:
There is no signature change to dev server, but for the dev middleware you are right - it could break usage (by the way: async middleware is described in [express docs](https://expressjs.com/en/guide/writing-middleware.html)).

Another way to solve that could be to resolve config in [ConfigLoader](https://github.com/PepsRyuu/nollup/blob/master/lib/impl/ConfigLoader.js), what do you think?

```js
// return config;
return (typeof config === 'function')
    ? await config()
    : config;
```

PepsRyuu commented:
I think that would work a lot better. Fits in well with the existing code and allows both the CLI and the dev-server to avail for function configs, which I think satisfies the majority of use cases.

piotr-cz commented:
OK, done: reverted and implemented in 74c48e2

I was trying to add new test cases but I haven't found any that utilize cli or dev-server, just directly nollup which accepts only objects.

Good work with tests by the way 👍

PepsRyuu commented:
Released in ```0.12.2``` :)
